File: ios/RezumateNative/Services/JobFitExtractor.swift
import Foundation

struct JobFit: Equatable {
    var jobTitle: String?
    var jobTitleMatched: Bool
    var educationRequirement: String?
    var educationMatched: Bool
}

enum JobFitExtractor {
    static func extract(jobDescription: String, resumeText: String) -> JobFit {
        let title = extractJobTitle(from: jobDescription)
        let education = extractEducationRequirement(from: jobDescription)
        let resumeLower = resumeText.lowercased()

        return JobFit(
            jobTitle: title,
            jobTitleMatched: title.map { resumeContainsTitle($0, resumeLower: resumeLower) } ?? false,
            educationRequirement: education,
            educationMatched: education.map { resumeContainsEducation($0, resumeLower: resumeLower) } ?? false
        )
    }

    static func extractJobTitle(from jobDescription: String) -> String? {
        let labeled = firstCapture(
            in: jobDescription,
            pattern: #"(?i)(?:job\s+title|title|role|position)\s*[:\-]\s*([^\n]{4,80})"#
        )
        if let labeled = cleanupTitle(labeled) {
            return labeled
        }

        let hiring = firstCapture(
            in: jobDescription,
            pattern: #"(?i)(?:we are )?(?:hiring|seeking|looking for)\s+(?:an?\s+)?([^.\n]{4,80})"#
        )
        return cleanupTitle(hiring)
    }

    static func extractEducationRequirement(from jobDescription: String) -> String? {
        let lower = jobDescription.lowercased()
        let checks: [(label: String, pattern: String)] = [
            ("PhD", #"\bph\.?d\b|\bdoctorate\b"#),
            ("MBA", #"\bmba\b"#),
            ("Master's degree", #"\bmaster'?s\b|\bm\.?s\.?\b|\bm\.?a\.?\b"#),
            ("Bachelor's degree", #"\bbachelor'?s\b|\bb\.?s\.?\b|\bb\.?a\.?\b|\bbachelor\b"#),
            ("Associate degree", #"\bassociate(?:'s)?\s+degree\b"#),
        ]

        for check in checks {
            if matches(check.pattern, in: lower) {
                return check.label
            }
        }
        return nil
    }

    private static func resumeContainsTitle(_ title: String, resumeLower: String) -> Bool {
        let lower = title.lowercased()
        if resumeLower.contains(lower) {
            return true
        }

        let tokens = lower
            .components(separatedBy: CharacterSet.alphanumerics.inverted)
            .filter { $0.count >= 4 && !titleNoise.contains($0) }
        guard tokens.count >= 2 else {
            return tokens.contains(where: { resumeLower.contains($0) })
        }
        let hits = tokens.filter { resumeLower.contains($0) }.count
        return hits >= max(2, tokens.count - 1)
    }

    private static func resumeContainsEducation(_ requirement: String, resumeLower: String) -> Bool {
        switch requirement {
        case "PhD":
            return matches(#"\bph\.?d\b|\bdoctorate\b"#, in: resumeLower)
        case "MBA":
            return matches(#"\bmba\b"#, in: resumeLower)
        case "Master's degree":
            return matches(#"\bmaster'?s\b|\bm\.?s\.?\b|\bm\.?sc\b"#, in: resumeLower)
        case "Bachelor's degree":
            return matches(#"\bbachelor'?s\b|\bb\.?s\.?\b|\bb\.?a\.?\b|\bb\.s\b|\bbachelor\b"#, in: resumeLower)
        case "Associate degree":
            return matches(#"\bassociate(?:'s)?\b"#, in: resumeLower)
        default:
            return resumeLower.contains(requirement.lowercased())
        }
    }

    private static func cleanupTitle(_ raw: String?) -> String? {
        guard var title = raw?.trimmingCharacters(in: .whitespacesAndNewlines), !title.isEmpty else {
            return nil
        }

        let cutTokens = [" to ", " who ", " that ", " with ", " for ", " in ", ","]
        let lowered = title.lowercased()
        var earliest: String.Index?
        for token in cutTokens {
            if let range = lowered.range(of: token) {
                if earliest == nil || range.lowerBound < earliest! {
                    earliest = range.lowerBound
                }
            }
        }
        if let earliest {
            title = String(title[..<earliest])
        }

        title = title.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines.union(.punctuationCharacters))
        let words = title.components(separatedBy: .whitespaces).filter { !$0.isEmpty }
        guard (2...8).contains(words.count) else { return nil }
        guard !titleNoise.contains(title.lowercased()) else { return nil }
        return title
    }

    private static let titleNoise: Set<String> = [
        "team", "someone", "candidate", "candidates", "person", "people", "individual", "role"
    ]

    private static func firstCapture(in text: String, pattern: String) -> String? {
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return nil }
        let range = NSRange(text.startIndex..., in: text)
        guard let match = regex.firstMatch(in: text, options: [], range: range),
              match.numberOfRanges > 1,
              let capture = Range(match.range(at: 1), in: text) else {
            return nil
        }
        return String(text[capture])
    }

    private static func matches(_ pattern: String, in text: String) -> Bool {
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return false }
        let range = NSRange(text.startIndex..., in: text)
        return regex.firstMatch(in: text, options: [], range: range) != nil
    }
}

File: android/app/src/main/java/com/aftaab/rezumate/domain/JobFitExtractor.kt
package com.aftaab.rezumate.domain

data class JobFit(
    val jobTitle: String? = null,
    val jobTitleMatched: Boolean = false,
    val educationRequirement: String? = null,
    val educationMatched: Boolean = false,
)

object JobFitExtractor {
    fun extract(jobDescription: String, resumeText: String): JobFit {
        val title = extractJobTitle(jobDescription)
        val education = extractEducationRequirement(jobDescription)
        val resumeLower = resumeText.lowercase()
        return JobFit(
            jobTitle = title,
            jobTitleMatched = title?.let { resumeContainsTitle(it, resumeLower) } ?: false,
            educationRequirement = education,
            educationMatched = education?.let { resumeContainsEducation(it, resumeLower) } ?: false,
        )
    }

    fun extractJobTitle(jobDescription: String): String? {
        labeledTitle(jobDescription)?.let { return it }
        return cleanupTitle(
            firstCapture(
                jobDescription,
                Regex(
                    """(?i)(?:we are )?(?:hiring|seeking|looking for)\s+(?:an?\s+)?([^.\n]{4,80})""",
                ),
            ),
        )
    }

    fun extractEducationRequirement(jobDescription: String): String? {
        val lower = jobDescription.lowercase()
        val checks = listOf(
            "PhD" to Regex("""\bph\.?d\b|\bdoctorate\b"""),
            "MBA" to Regex("""\bmba\b"""),
            "Master's degree" to Regex("""\bmaster'?s\b|\bm\.?s\.?\b|\bm\.?a\.?\b"""),
            "Bachelor's degree" to Regex("""\bbachelor'?s\b|\bb\.?s\.?\b|\bb\.?a\.?\b|\bbachelor\b"""),
            "Associate degree" to Regex("""\bassociate(?:'s)?\s+degree\b"""),
        )
        return checks.firstOrNull { it.second.containsMatchIn(lower) }?.first
    }

    private fun labeledTitle(jobDescription: String): String? = cleanupTitle(
        firstCapture(
            jobDescription,
            Regex("""(?i)(?:job\s+title|title|role|position)\s*[:\-]\s*([^\n]{4,80})"""),
        ),
    )

    private fun resumeContainsTitle(title: String, resumeLower: String): Boolean {
        val lower = title.lowercase()
        if (lower in resumeLower) return true
        val tokens = lower
            .split(Regex("[^a-z0-9]+"))
            .filter { it.length >= 4 && it !in TITLE_NOISE }
        if (tokens.size < 2) return tokens.any { it in resumeLower }
        val hits = tokens.count { it in resumeLower }
        return hits >= maxOf(2, tokens.size - 1)
    }

    private fun resumeContainsEducation(requirement: String, resumeLower: String): Boolean = when (requirement) {
        "PhD" -> Regex("""\bph\.?d\b|\bdoctorate\b""").containsMatchIn(resumeLower)
        "MBA" -> Regex("""\bmba\b""").containsMatchIn(resumeLower)
        "Master's degree" -> Regex("""\bmaster'?s\b|\bm\.?s\.?\b|\bm\.?sc\b""").containsMatchIn(resumeLower)
        "Bachelor's degree" ->
            Regex("""\bbachelor'?s\b|\bb\.?s\.?\b|\bb\.?a\.?\b|\bb\.s\b|\bbachelor\b""").containsMatchIn(resumeLower)
        "Associate degree" -> Regex("""\bassociate(?:'s)?\b""").containsMatchIn(resumeLower)
        else -> requirement.lowercase() in resumeLower
    }

    private fun cleanupTitle(raw: String?): String? {
        var title = raw?.trim().orEmpty()
        if (title.isEmpty()) return null

        val lowered = title.lowercase()
        val cutTokens = listOf(" to ", " who ", " that ", " with ", " for ", " in ", ",")
        val earliest = cutTokens
            .mapNotNull { token -> lowered.indexOf(token).takeIf { it >= 0 } }
            .minOrNull()
        if (earliest != null) title = title.substring(0, earliest)

        title = title.trim().trim { it.isWhitespace() || it in PUNCTUATION }
        val words = title.split(Regex("\\s+")).filter(String::isNotEmpty)
        if (words.size !in 2..8) return null
        if (title.lowercase() in TITLE_NOISE) return null
        return title
    }

    private fun firstCapture(text: String, regex: Regex): String? =
        regex.find(text)?.groupValues?.getOrNull(1)

    private val TITLE_NOISE = setOf(
        "team", "someone", "candidate", "candidates", "person", "people", "individual", "role",
    )

    private val PUNCTUATION = setOf(
        ',', '.', '!', '?', ';', ':', '"', '\'', ')', '(', '[', ']', '{', '}', '-', '—', '–',
    )
}
